### PR TITLE
fix PodLogsQuerySplitStream if feature is enabled and using defaults

### LINKED ISSUE #128709: If one enables PodLogsQuerySplitStreams and aims to access logs without stream, a validation error occurs
### What happened?

We saw this in our alpha jobs.

https://prow.k8s.io/view/gs/kubernetes-ci-logs/logs/ci-kubernetes-e2e-kind-alpha-features/1854662086911594496

https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/128680/pull-e2e-gci-gce-alpha-enabled-default/1854616736976867328

### What did you expect to happen?

One should be able to run the following command with this feature gate on:

```
kehannon@kehannon-thinkpadp1gen4i:~/Work/KubeExamples$ k get --raw /api/v1/namespaces/default/pods/example/log
The PodLogOptions "example" is invalid: stream: Required value: must be specified
```

### How can we reproduce it (as minimally and precisely as possible)?

Start a cluster with PodLogsQuerySplitStreams enabled.

### Anything else we need to know?

_No response_

### Kubernetes version

1.32

### Cloud provider

NA

### OS version

_No response_

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

### PR DESCRIPTION
**Pull request by kannon92**
#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:
When using the Rest API for Pod Logs, we need to default to All if the feature gate is turned on.

Otherwise, all log calls will fail with a validation error.
#### Which issue(s) this PR fixes:

Fixes #128709 

#### Special notes for your reviewer:

One could reproduce this by creating a pod and use the raw rest api:

```
kehannon@kehannon-thinkpadp1gen4i:~/Work/KubeExamples$ k get --raw /api/v1/namespaces/default/pods/example/log
The PodLogOptions "example" is invalid: stream: Required value: must be specified
```
I can confirm that this does not error once this change correctly defaults to All if feature is on.

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by liggitt** [comment]
I'll defer to @thockin as API reviewer for https://github.com/kubernetes/kubernetes/pull/127360

looks like this out-of-band defaulting is already being done in the kubelet log handling prior to validation ... duplicating this here seems necessary, I guess

looks like normal defaulting to params objects doesn't get applied if there are no query params... that's a sharp edge:

https://github.com/kubernetes/kubernetes/blob/591c75e40b0c861b35e69d4818dffb75f9668402/staging/src/k8s.io/apimachinery/pkg/runtime/codec.go#L152-L185

File: pkg/registry/core/pod/rest/log.go [hunk 1]
```diff
@@ -90,6 +91,11 @@ func (r *LogREST) Get(ctx context.Context, name string, opts runtime.Object) (ru
 
 	if !utilfeature.DefaultFeatureGate.Enabled(features.PodLogsQuerySplitStreams) {
 		logOpts.Stream = nil
+	} else {
```
Comment (thockin) [inline pkg/registry/core/pod/rest/log.go:94]: This is not right.  If the gate is enabled, defaults should have been set. 

pkg/apis/core/v1/defaults.go:
```
442 func SetDefaults_PodLogOptions(obj *v1.PodLogOptions) {                                                                          
443     if utilfeature.DefaultFeatureGate.Enabled(features.PodLogsQuerySplitStreams) {
444         if obj.Stream == nil {
445             obj.Stream = ptr.To(v1.LogStreamAll)
446         }
447     }
448 }
```

and

pkg/apis/core/v1/zz_generated.defaults.go:
```
51     scheme.AddTypeDefaultingFunc(&corev1.PodLogOptions{}, func(obj interface{}) { SetObjectDefaults_PodLogOptions(obj.(*corev1.P     odLogOptions)) })
```

So why is it not being called?

Comment (liggitt) [reply pkg/registry/core/pod/rest/log.go:94]: > normal defaulting to params objects doesn't get applied if there are no query params

https://github.com/kubernetes/kubernetes/pull/128712#issuecomment-2465809232

**Comment by thockin** [comment]
Ugh, I flagged that in the kubelet server path, but this is apiserver.  I just assumed that we passed the PodLogOptions as the HTTP body.

Joe is already aware of this as a "sharp edge".  https://github.com/kubernetes/kubernetes/issues/128589

File: pkg/registry/core/pod/rest/log.go [hunk 2]
```diff
@@ -90,6 +91,11 @@ func (r *LogREST) Get(ctx context.Context, name string, opts runtime.Object) (ru
 
 	if !utilfeature.DefaultFeatureGate.Enabled(features.PodLogsQuerySplitStreams) {
```
Comment (thockin) [inline pkg/registry/core/pod/rest/log.go:92]: Do we prefer this or just to "fix" validation to not assert Required when nil?  We already have this:

pkg/kubelet/server/server.go:
```
 727     if utilfeature.DefaultFeatureGate.Enabled(features.PodLogsQuerySplitStreams) {                                              
 728         // Even with defaulters, logOptions.Stream can be nil if no arguments are provided at all.
 729         if logOptions.Stream == nil {
 730             // Default to "All" to maintain backward compatibility.
 731             logOptions.Stream = ptr.To(v1.LogStreamAll)
 732         } 
 733     }
```

I think that's a little bit cleaner?

**Comment by knight42** [comment]
Thanks for the quick fix!

File: pkg/registry/core/pod/rest/log.go (hunk 2 above)
Comment (kannon92) [reply pkg/registry/core/pod/rest/log.go]: Yea, I like this. Went ahead and added this.

**Review by thockin — COMMENTED** [review]
That's not really what I meant. But on considering, it's fine.  I don't love it, but hopefully we fix the decode issue and remove this.

/approve
/lgtm

**Pull request merged**